deepinv/sampling/samplers.py

```diff
@@ -0,0 +1,344 @@
+import sys
+from collections import deque
+import torch
+from tqdm import tqdm
+from deepinv.physics import Physics
+from deepinv.optim.optim_iterators import *
+from deepinv.optim.prior import Prior
+from deepinv.models import Reconstructor
+from deepinv.optim.data_fidelity import DataFidelity
+from deepinv.sampling.sampling_iterators.sample_iterator import SamplingIterator
+from deepinv.sampling.utils import Welford, projbox
+from deepinv.sampling.sampling_iterators import *
+from deepinv.optim.utils import check_conv
+
+
+class BaseSample(Reconstructor):
+    r"""
+    Base class for Monte Carlo sampling.
+
+    This class can be used to create new Monte Carlo samplers by implementing the sampling kernel through :class:`deepinv.sampling.SamplingIterator`:
+
+    ::
+
+        # define your sampler (possibly a Markov kernel which depends on the previous sample)
+        class MyIterator(SamplingIterator):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y, physics, data_fidelity, prior, params_algo):
+                # run one sampling kernel iteration
+                new_x = f(x, y, physics, data_fidelity, prior, params_algo)
+                return new_x
+
+        # create the sampler
+        sampler = BaseSampler(MyIterator(), prior, data_fidelity, iterator_params)
+
+        # compute posterior mean and variance of reconstruction of x
+        mean, var = sampler(y, physics)
+
+    This class computes the mean and variance of the chain using Welford's algorithm, which avoids storing the whole
+    Monte Carlo samples. It can also maintain a history of the `history_size` most recent samples.
+
```

Comment:
Given that we do not run until convergence, but for a fixed number of iterations, and that the final number of samples computed is not directly computed: To improve clarity, could you add a note/comment here with a small example saying how many samples are computed for a given `max_iter`, `burnin_rate` and `thinning` values? For example taking the default values of each parameter.